File: LeetCode/22.php
<?php

class Solution
{

    private $result = [];

    /**
     * @param Integer $n
     * @return String[]
     */
    function generateParenthesis($n)
    {
        $this->name('(', 1, 0, $n);
        return $this->result;
    }

    private function name(string $str, int $left, int $right, int $n)
    {
        if ($left > $n || $right > $n || $right > $left) return;

        if ($left == $right && $left == $n) {
            return $this->result[] = $str;
        }

        $this->name($str . '(', $left + 1, $right, $n);
        $this->name($str . ')', $left, $right + 1, $n);
    }
}


// TODO:test
$test = new solution();
var_dump($test->generateParenthesis(3));

File: LeetCode/22.java
import java.util.*;

class Solution {
    public List<String> generateParenthesis(int n) {
        List<String> result = new ArrayList<String>();
        helper(result, "", 0, 0, n);
        return result;
    }

    private void helper(List<String> result, String str, int left, int right, int n) {
        if (left > n || right > n || right > left) {
            return;
        }
        if (left == right && left == n) {
            result.add(str);
        }
        helper(result, str + "(", left + 1, right, n);
        helper(result, str + ")", left, right + 1, n);
    }
}
